File: Scr_Java/modelo/Coordenada.java
package modelo;

import java.util.Objects;

public class Coordenada {
    private final int fila;
    private final int columna;

    public Coordenada(int fila, int columna) {
        this.fila = fila;
        this.columna = columna;
    }

    public int getFila() {
        return fila;
    }

    public int getColumna() {
        return columna;
    }

    // Distancia Manhattan: se usa para decidir si un enemigo esta dentro
    // del alcance de una torre (Torre.estaEnRango).
    public int distanciaA(Coordenada otra) {
        return Math.abs(this.fila - otra.fila) + Math.abs(this.columna - otra.columna);
    }

    // equals/hashCode son necesarios porque Coordenada se usa como clave logica
    // para comparar posiciones: ocupacion de casillas en Mapa y busqueda de torres.
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Coordenada)) return false;
        Coordenada that = (Coordenada) o;
        return fila == that.fila && columna == that.columna;
    }

    @Override
    public int hashCode() {
        return Objects.hash(fila, columna);
    }

    @Override
    public String toString() {
        return "(" + fila + ", " + columna + ")";
    }
}

File: cpp/modelo/Coordenada.cpp
#include "Coordenada.h"

#include <cmath>
#include <sstream>

// Constructor de la coordenada.
Coordenada::Coordenada(int fila, int columna)
    : fila(fila), columna(columna) {
}

// Retorna la fila de la coordenada.
int Coordenada::getFila() const {
    return fila;
}

// Retorna la columna de la coordenada.
int Coordenada::getColumna() const {
    return columna;
}

// Distancia Manhattan: se usa para decidir si un enemigo esta dentro
// del alcance de una torre (Torre.estaEnRango).
int Coordenada::distanciaA(const Coordenada& otra) const {
    return std::abs(fila - otra.fila) +
           std::abs(columna - otra.columna);
}

// Compara dos coordenadas para determinar si representan
// la misma posicion en el mapa.
bool Coordenada::operator==(const Coordenada& otra) const {
    return fila == otra.fila &&
           columna == otra.columna;
}

// Retorna la coordenada en formato de texto.
std::string Coordenada::toString() const {
    std::ostringstream salida;

    salida << "(" << fila << ", " << columna << ")";

    return salida.str();
}